address_book/src/lib.rs

```diff
@@ -32,15 +29,48 @@ static LOGGING: Lazy<()> = Lazy::new(|| unsafe {
     tide::log::with_level(LOG_LEVEL);
 });
 
+trait Store {
+    fn save(&self, address: UserAddress, pubkey: UserPubKey) -> Result<(), std::io::Error>;
+    fn load(&self, address: UserAddress) -> Option<UserPubKey>;
+}
+
+#[derive(Debug, Clone)]
+struct FileStore {
+    dir: PathBuf,
+}
+
+impl FileStore {
+    fn new(dir: PathBuf) -> Self {
+        Self { dir }
+    }
+    fn path(&self, address: UserAddress) -> PathBuf {
+        let as_hex = hex::encode(bincode::serialize(&address).unwrap());
+        self.dir.join(format!("{}.bin", as_hex))
+    }
+}
+
+impl Store for FileStore {
+    fn save(&self, address: UserAddress, pubkey: UserPubKey) -> Result<(), std::io::Error> {
```

Comment:
From the documentation, I'm not seeing anything guaranteeing this is incapable of a mid-write flush with subsequent read collision under NFS-over-EFS.

It would, however, be guaranteed if you did the fs::write to a path with a temp suffix (preferably with a random component) and then did a fs::rename after the write.

Reply:
Thanks. Yeah so far I haven't been able to figure out what kind of behavior we would see in practice. I think we're unlikely to run into any issues for the demo but I don't mind adding the tempfile/rename solution. Although one wonders if we shouldn't just use S3.

Reply:
This may be relevant https://stackoverflow.com/a/11402827 but no link to any documentation. It suggests that if we stay below the block size the files should remain consistent.

Reply:
Sure, or Dynamo, but both of those require using an SDK, with the learning and deployment overhead that comes with that, both are AWS lock-in, and testing isn't quite as transparently symmetrical. This solution has the benefit of being both simple and trivial to test locally, and it works on any cloud service that provides a distributed filesystem that can map to a unix-like file system (e.g. Azure or Google), which may not matter that much, but is still nice.

I know from prior experience that EFS updates are fast enough that it's hard to calibrate the delay across nodes vs delay from other factors (network access to the node access point, etc.), and that the propagation time for remotely created files was small enough that it got buried in the network noise for a native (C++) testing framework pinging its counterpart over a direct network socket.

Reply:
Oh, right, I forgot about the NFS negotiated block message. You actually get to set the block size when you implement your NFS adaptor, but there's a bunch of stuff about unstable writes, and I think AWS is using the v4 standard. It's been a long time since I've dealt with this, and nothing had actually implemented v4 last time I had to muck with the guts of NFS. I'm making a note to myself to do some reading when I've got some time.

For what it's worth, this is the AWS documentation: https://docs.aws.amazon.com/efs/latest/ug/mounting-fs-old.html#mounting-fs-nfs-info and current NFS documentation: https://datatracker.ietf.org/doc/html/rfc3530.txt (4.0) and https://datatracker.ietf.org/doc/html/rfc5661 (4.1)

The default mount is in NFS 4.1 mode. It looks like their default configured block size is actually 1MB (https://docs.aws.amazon.com/efs/latest/ug/efs-mount-helper.html#efs-mount-helper-how - see the rsize/wsize settings). So, yeah, even without the rename, this _should_ be fine. The rename pattern is, however, best practice for a reason, and I'd prefer its use.